modules/service/src/test/scala/lucuma/odb/graphql/issue/shortcut/5017.scala

```diff
@@ -0,0 +1,138 @@
+        _ <- addStepEventAs(serviceUser, s, StepStage.EndStep)
+      yield d
+
+    // Execute an arc, a flat and one science step
+    def recordAndExecuteScienceStep(c: AtomicCell[IO, Int], a: Atom.Id, ditherNm: Int, q: Int): IO[(Step.Id, Dataset.Id)] =
+      for
+        s <- recordStepAs(serviceUser, a, Instrument.GmosNorth, gmosNorthScience(ditherNm), StepConfig.Science, sciTelescopeConfig(q), ObserveClass.Science)
+        d <- executeStep(c, s)
+      yield (s, d)
+
+    def executeAtom(c: AtomicCell[IO, Int], v: Visit.Id, ditherNm: Int, q: Int, stepCount: Int): IO[(Atom.Id, List[(Step.Id, Dataset.Id)])] =
+      for
+        a  <- recordAtomAs(serviceUser, Instrument.GmosNorth, v, SequenceType.Science)
+        sa <- recordStepAs(serviceUser, a, Instrument.GmosNorth, gmosNorthArc(ditherNm), ArcStep, gcalTelescopeConfig(q), ObserveClass.PartnerCal)
+        da <- executeStep(c, sa)
+        sf <- recordStepAs(serviceUser, a, Instrument.GmosNorth, gmosNorthFlat(ditherNm), FlatStep, gcalTelescopeConfig(q), ObserveClass.PartnerCal)
+        df <- executeStep(c, sf)
+        ss <- recordAndExecuteScienceStep(c, a, ditherNm, q).replicateA(stepCount)
+      yield (a, (sa, da) :: (sf, df) :: ss)
+
+    val setup: IO[(Observation.Id, (Atom.Id, List[(Step.Id, Dataset.Id)]))] =
+      for
+        p <- createProgram
+        _ <- setProgramReference(staff, p, """engineering: { semester: "2025B", instrument: GMOS_NORTH }""")
+        t <- createTargetWithProfileAs(pi, p)
+        o <- createGmosNorthLongSlitObservationAs(pi, p, List(t))
+        v <- recordVisitAs(serviceUser, Instrument.GmosNorth, o)
+        c <- AtomicCell[IO].of(0)
+        _ <- executeAtom(c, v,  0,   0, 3).void
+        _ <- executeAtom(c, v,  5,  15, 3).void
+        _ <- executeAtom(c, v, -5, -15, 3).void
+        a <- executeAtom(c, v,  0,   0, 1)
+      yield (o, a)
+
+    def currentStateQuery(o: Observation.Id): String = s"""
+      query {
+        observation(observationId: "$o") {
+          execution {
+            digest {
+              science {
+                timeEstimate {
+                  total { microseconds }
+                }
+              }
+            }
+          }
+          workflow { state }
+        }
+      }
+    """
+
+    // workflow state and remaining time (microseconds)
+    def currentState(o: Observation.Id): IO[(ObservationWorkflowState, Long)] =
+      query(
+        user = pi,
+        query = currentStateQuery(o)
+      ).flatMap: json =>
+        val c = json.hcursor.downField("observation")
+        (for
+          s <- c.downFields("workflow", "state").as[ObservationWorkflowState]
+          t <- c.downFields("execution", "digest", "science", "timeEstimate", "total", "microseconds").as[Long]
+        yield (s, t)).leftMap(f => new RuntimeException(f.message)).liftTo[IO]
+
+    val obtained =
+      for
+        init     <- setup
+        executed <- currentState(init._1)
```

Comment:
Delaying for a second before sampling the state above seems to make it work reliably.

Reply:
urgh

Reply:
When it fails, it is because the ITC results are not present.

Reply:
And furthermore, it isn't that the ITC results have been deleted or that the hash no longer matches or something but rather they just haven't yet been written at all and only appear later after the workflow state has already been requested.  Or so it seems.

Reply:
Is there any kind of subscription event sent when ITC results are updated? I thought there was, but I could be thinking of something else.

Reply:
Yes, when the digest or itc cache result is updated there is an observation edit event.  I don't think that's related?

Reply:
I would think explore should re-request the workflow when it gets the observation edit event for the itc cache update and the workflow would then match.

Reply:
Hmm, yes, perhaps the nondeterministic part of this comes from the fact that I request the digest and the workflow state in the same query.  The workflow is calculated with an effect handler.  If I explicitly request the digest first and then ask for the workflow it seems to work.